ycmd/tests/get_completions_test.py

```diff
@@ -115,9 +123,800 @@ def GetCompletions_IdentifierCompleter_WorksForSpecialIdentifierChars_test():
                            completion_data ).json[ 'completions' ]
 
   assert_that( results,
-               has_items( CompletionEntryMatcher( 'font-size' ),
-                          CompletionEntryMatcher( 'font-family' ) ) )
+               has_items( CompletionEntryMatcher( 'font-size', '[ID]' ),
+                          CompletionEntryMatcher( 'font-family', '[ID]' ) ) )
+
 
+@with_setup( Setup )
+def GetCompletions_Fall_Back_ClangCompleter_test():
+
+  # because we use it a lot
+  NO_COMPLETIONS_ERROR = ErrorMatcher( RuntimeError, NO_COMPLETIONS_MESSAGE )
+
+  """
+  GetCompletions_Fall_Back_test
+
+  Tests for:
+    - [X] multiple languages: (use cases)
+      - [X] C:      macros
+      - [X] C++:    templates
+      - [X] Python: properties set from external method
+    - [X] Forced vs. Triggered semantic completion
+    - [X] Identifiers tagged with '[ID]'
+    - [X] Filter/sort based on query
+    - [X] Min chars for completion (only applies to ID)
+    - [X] Extra conf throws exception
```

Comment:
I have a feeling you're testing too many things at once.

For instance, I'd have just one test that tests the fallback (for say C++). The codepath that implements the fallback is agnostic of the semantic engine, so testing that the fallback works for various semantic engines seems overkill to me.

Likewise, for semantic engine tests, just use force_semantic to not have to test the fallback logic. That's tested in the fallback test.

A good design idea for tests is that if one logical thing fails in the code, one test should fail. That's often not achievable, but something we should strive for. The idea is that if one logical component fails and 100 tests fail because of it, tracking down exactly which component failed because non-trivial.